core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java

```diff
@@ -378,21 +405,49 @@ public void addLocation(long id, BlockLocation location) {
   @Override
   public void removeLocation(long blockId, long workerId) {
     byte[] key = RocksUtils.toByteArray(blockId, workerId);
-    try {
+    try (LockResource lock = checkAndAcquireReadLock()) {
       db().delete(mBlockLocationsColumn.get(), mDisableWAL, key);
     } catch (RocksDBException e) {
       throw new RuntimeException(e);
     }
   }
 
   @Override
+  /**
+   * This iterator is used in:
+   * 1. {@link BlockIntegrityChecker} to iterate all existing blocks
+   * 2. Journal dumping like checkpoint/backup sequences
+   */
   public CloseableIterator<Block> getCloseableIterator() {
-    RocksIterator iterator = db().newIterator(mBlockMetaColumn.get(), mIteratorOption);
-    return RocksUtils.createCloseableIterator(iterator,
-        (iter) -> new Block(Longs.fromByteArray(iter.key()), BlockMeta.parseFrom(iter.value())));
+    try (LockResource lock = checkAndAcquireReadLock()) {
+      RocksIterator iterator = db().newIterator(mBlockMetaColumn.get(), mIteratorOption);
+      return RocksUtils.createCloseableIterator(iterator,
+          (iter) -> new Block(Longs.fromByteArray(iter.key()), BlockMeta.parseFrom(iter.value())),
+          mClosed::get, this::checkAndAcquireReadLock);
+    }
   }
 
   private RocksDB db() {
     return mRocksStore.getDb();
   }
+
+  private void checkDbStatus() {
+    if (mClosed.get()) {
+      throw new UnavailableRuntimeException(
+          "RocksDB is closed. Master is failing over or shutting down.");
+    }
+  }
+
+  // TODO(jiacheng): double check what happens if max lock count error here
+  public LockResource checkAndAcquireReadLock() {
```

Comment:
Currently this is what I'm not sure about. 

On a max lock count error, a `java.lang.Error` will be thrown and the thread will die. But most of these calls are from RPC threads and dead RPC threads don't crash the process. What's the best thing to do here? If I just catch Error and `ProcessUtils.fatalError()`, would that be an overkill?